File: engine/camera/Camera.go
package camera

import (
	"github.com/go-gl/mathgl/mgl32"
)

const (
	FORWARD  int = 0
	BACKWARD int = 1
	LEFT         = 2
	RIGHT        = 3
)

// Default camera values
const (
	YAW         float32 = -90.0
	PITCH       float32 = -45.0
	SPEED               = 0.025
	SENSITIVITY         = 0.1
	ZOOM                = 45.0
)

type Camera struct {
	// camera attributes
	Position mgl32.Vec3
	Target   mgl32.Vec3

	Front   mgl32.Vec3
	Right   mgl32.Vec3
	Up      mgl32.Vec3
	WorldUp mgl32.Vec3

	// camera options
	MovementSpeed    float32
	MouseSensitivity float32
	Zoom             float32
}

func (c *Camera) Init(position mgl32.Vec3, target mgl32.Vec3) {
	c.Position = position
	c.Target = target

	c.Front = c.Target.Sub(c.Position).Normalize()
	c.Up = mgl32.Vec3{0.0, 1.0, 0.0}
	c.WorldUp = mgl32.Vec3{0.0, 1.0, 0.0}
	c.MovementSpeed = SPEED

	c.Zoom = ZOOM
}

func (c *Camera) GetViewMatrix() mgl32.Mat4 {
	return mgl32.LookAtV(c.Position, c.Target, c.Up)
}

func (c *Camera) ProcessKeyboard(direction int, deltaTime float64) {
	c.Front = c.Target.Sub(c.Position).Normalize()

	velocity := c.MovementSpeed * float32(deltaTime)
	switch direction {
	case FORWARD:
		c.Position = mgl32.HomogRotate3DX(-velocity).Mul4x1(c.Position.Vec4(0)).Vec3()
	case BACKWARD:
		c.Position = mgl32.HomogRotate3DX(+velocity).Mul4x1(c.Position.Vec4(0)).Vec3()
	case LEFT:
		c.Position = mgl32.HomogRotate3DY(-0.1).Mul4x1(c.Position.Vec4(0)).Vec3()
	case RIGHT:
		c.Position = mgl32.HomogRotate3DY(+0.1).Mul4x1(c.Position.Vec4(0)).Vec3()
	}
}

func (c *Camera) ProcessMouseScroll(yOffset float32) {
	c.Zoom -= yOffset
	if c.Zoom < 1.0 {
		c.Zoom = 1.0
	}
	if c.Zoom > 45.0 {
		c.Zoom = 45.0
	}
}

File: engine/camera/camera.cpp
#include "camera.h"
#include <glm/gtc/matrix_transform.hpp>

#include <cmath>

/*
 * Camera 实现（状态容器版本）
 *
 * 旧的 Camera 同时承载了轨道/平移/缩放等输入逻辑（OrbitAroundOrigin/Pan/Zoom）
 * 与自由第一人称逻辑（ProcessKeyboard/ProcessMouseMovement），两类逻辑互相
 * 覆盖同一组 front/right/up 状态，且后者从未被调用。重构后相机只剩状态与
 * 视图矩阵生成，交互逻辑移入 OrbitManipulator（见 manipulator.h）。
 */

// 默认构造：位置与朝向留默认值（位置原点、朝向单位四元数 = 看向 -Z）
Camera::Camera()
{
}

/*
 * 参数化构造：以「位置-目标-世界上方向」初始化相机
 *
 * 与旧版语义一致，用于从 CameraConfig 加载摄像机：
 *   - position 相机位置
 *   - target   初始注视目标（用于推导朝向）
 *   - world_up 世界上方向
 * 内部把"朝向 = 目标 - 位置"换算为四元数存储（见 LookAt）。
 */
Camera::Camera(glm::vec3 position, glm::vec3 target, glm::vec3 world_up)
    : m_position(position)
{
    LookAt(target, world_up);
}

Camera::~Camera()
{
}

std::string Camera::GetName() const {
    return m_name;
}

// 移动相机位置（纯平移，不改变朝向；操控器在需要时自行调整朝向）
void Camera::SetPosition(const glm::vec3 &position) {
    m_position = position;
}

// 由操控器写入朝向四元数（约定：相机前方 = 四元数旋转后的 -Z）
void Camera::SetOrientation(const glm::quat &orientation) {
    m_orientation = orientation;
}

/*
 * 以「目标点 + 世界上方向」重建朝向
 *
 * 依据 position 与 target 推导视线方向 dir，再用 dir 与 world_up 构造
 * 四元数：glm::quatLookAt(dir, up) 生成使相机前方（-Z）正对 dir 的旋转。
 * 视线与世界上方向几乎平行时（往正上/正下看）改为使用 Z 轴作为上方向，
 * 避免叉积退化为零向量导致 NaN（与旧 OrbitAroundOrigin 的 ±89° 限制同源）。
 */
void Camera::LookAt(const glm::vec3 &target, const glm::vec3 &world_up) {
    const glm::vec3 dir = glm::normalize(target - m_position);

    // 视线与 up 平行时退化：改用 Z 轴充当上方向
    const glm::vec3 safeUp = (std::fabs(glm::dot(dir, world_up)) > 0.999f)
                                 ? glm::vec3(0.0f, 0.0f, 1.0f)
                                 : world_up;
    m_orientation = glm::quatLookAt(dir, safeUp);
}

/*
 * 视图矩阵
 *
 * 由「位置 + 四元数朝向」构造 lookAt：
 *   - 相机前方 = 四元数旋转后的 -Z（OpenGL 相机默认看向 -Z）
 *   - 上方向   = 四元数旋转后的 +Y
 * 分别作为 lookAt 的注视点方向与上向量，保证朝向与矩阵严格一致。
 */
glm::mat4 Camera::GetViewMatrix() const {
    // 相机本地 -Z 轴经四元数旋转后即世界空间下的观察方向
    const glm::vec3 front = m_orientation * glm::vec3(0.0f, 0.0f, -1.0f);
    // 相机本地 +Y 轴经四元数旋转后即世界空间下的上方向
    const glm::vec3 up = m_orientation * glm::vec3(0.0f, 1.0f, 0.0f);
    return glm::lookAt(m_position, m_position + front, up);
}